### ISSUE TITLE
XENT loss function defined incorrectly?

### ISSUE THREAD

**Issue by ysmiraak** [deeplearning4j/nd4j#1169]
When using dl4j for binary classification, I want to use XENT as my loss function, but could never get the classifier to learn, so in the end I always turn to MSE or NEGATIVELOGLIKELIHOOD. And just now I checked the source code here:

https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/LossFunctions.java
and here
https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/LossCalculation.java

It seems that the return value is missing a negative sign. Could someone verify this?

Also, it seems to me that the definition for RECONSTRUCTION_CROSSENTROPY and XENT here are exactly the same. I couldn't find anywhere what RECONSTRUCTION_CROSSENTROPY actually means, it would be great if someone could explain this one to me.

**Comment by ysmiraak** [comment]
Moreover, I can't find any difference between NEGATIVELOGLIKELIHOOD and MCXENT.

In class LossCalculation, SQUARED_LOSS and MSE are also the same, and in class LossFunctions, the MSE loss is just the half of SQUARED_LOSS, which also confuses me: The mean should be dividing over the length of labels, right?

The error in RMSE_XENT is reported here: https://github.com/deeplearning4j/nd4j/issues/669
And it is still not fixed in 0.4.0.

I wonder if EXPLL is ok, but I don't understand anything about Poisson regression.

Maybe I am completely mistaken about loss functions? I'm a beginning learner, and this is quite frustrating to me. Any help would be great.

**Comment by eraly** [comment]
@ysmiraak 
So the error in RMSE: 
The RMSE is not something that is usually used as a loss function. You could in theory use anything as a loss function as long as the losses over all examples are averaged and the loss can be expressed as a function of the output of the net. We might add the RMSE to our list of functions but it is lower down than a lot of the others.

On if the negative log likelihood is missing a negative sign - 
It is not. If you traced through the code, you'd see it gets multiplied by -1 after it returns from the function in the same class here.
https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/LossCalculation.java#L72

The other questions -
The cross entropy is closely related to the maximum likelihood estimate. So these are very similar mathematical approaches - one about how close two probability distributions and the other about a statistical approach for estimating the parameters in a model given other values. 
In any case, the long and short of it is use mean squared error if your outputs are not one hot vectors. If your outputs are one hot vectors then you need to use the cross entropy function and make sure the activation function in your last layer is a softmax.

There are other loss functions other than these simplistic two. We are going through a rewrite and will post the documentation when the rewrite is ready.

Thanks!
